Issues 187 203 1059 command line improvements

FrankRay78 opened the pull request:
*** This PR contains the following PRs I've worked on - https://github.com/spectreconsole/spectre.console/pull/1101, https://github.com/spectreconsole/spectre.console/pull/1070, https://github.com/spectreconsole/spectre.console/pull/1102, rebased, merged together and with conflicts resolved ***

The following issues are addressed by the individual PRs listed above:

https://github.com/spectreconsole/spectre.console/issues/187, 'Allow default command configuration for a branch'
https://github.com/spectreconsole/spectre.console/issues/203, 'Simple switches end up on "remaining" list even if parsed properly'
https://github.com/spectreconsole/spectre.console/issues/1059, 'Ability to add command line flags and their values to remaining arguments if they cannot be set on the settings class'

Together they seem like a reasonable parcel of command line improvements, and so I've grouped them together here for ease of reviewing (should they wish to be released together).

patriksvensson commented:
@FrankRay78 This might seem odd, but could you send me an email at `patrik@patriksvensson.se`? I have a question to ask you in private. Normally I would check your git email, but it's anonymous.

patriksvensson commented on src/Spectre.Console.Cli/Internal/Modelling/CommandModelBuilder.cs:
```diff
@@ -1,5 +1,9 @@
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Spectre.Console.Cli.Tests")]
```
Is there a specific reason for this? I would prefer that we didn't expose internals if we don't really have to, even for tests. It makes them brittle.

Also: Assembly attributes should be placed in `Properties/AssemblyInfo.cs`

FrankRay78 replied:
I had the desire to unit test the CommandModelBuilder class and so wrote a bunch of tests here: https://github.com/FrankRay78/spectre.console/blob/Issues-187-203-1059-Command-line-improvements/test/Spectre.Console.Cli.Tests/Unit/CommandModelBuilderTests.cs, 

The only way I could work out how to expose the class for unit testing whilst keeping it inaccessible for regular spectre.console users, was to use the InternalsVisibleTo attribute, but restrict it to only the CommandModelBuilder class, hence why it's located in the class itself rather than `Properties/AssemblyInfo.cs`

FrankRay78 replied:
Fixed and branch updated.

patriksvensson commented on test/Spectre.Console.Cli.Tests/Unit/CommandModelBuilderTests.cs:
```diff
@@ -0,0 +1,210 @@
+namespace Spectre.Console.Cli.Tests.Unit;
+
+public sealed class CommandModelBuilderTests
```
Instead of exposing `CommandModel`, I think we should dump the command structure using the built-in XML command. You can see examples here: https://github.com/spectreconsole/spectre.console/blob/main/test/Spectre.Console.Cli.Tests/Unit/CommandAppTests.Xml.cs#L11

FrankRay78 replied:
I agree.  My use of the `InternalsVisibleTo` attribute is out of characteristic with the rest of the solution anyhow.  I'm not sure if the XML tests are new, and or I just didn't understand the codebase sufficiently when I was making these changes.  But I think they current CommandModelBuilder tests can be entirely replaced by the existing/adding some more xml tests.

FrankRay78 replied:
Fixed and branch updated.

patriksvensson commented on src/Spectre.Console.Cli/ICommandAppSettings.cs:
```diff
@@ -49,7 +49,15 @@ public interface ICommandAppSettings
+    /// during parsing with the message "Flags cannot be assigned a value."
+    /// should instead be added to the remaining arguments collection.
+    /// </summary>
+    bool ConvertFlagsToRemainingArgumentsIfCannotBeAssigned { get; set; }
```
Could we find a shorter name for this?

FrankRay78 replied:
Fixed and branch updated.

FrankRay78 commented:
Hi @patriksvensson, is there anything more required from myself regarding this PR?  I've had an initial pass regarding addressing all your review comments, and pushed these changes to the branch.  Let me know if you require anything further.
